### DISCUSSION TITLE
Patterns for accessing state in error path.

### DISCUSSION THREAD

**Discussion by askreet** [tokio-rs/axum#2616] [Q&A]
### Summary

I'm using a `Result<Response, AppError>` in order to make error paths quickly display an HTML error, like so:

```rust
struct AppError(anyhow::Error);

impl IntoResponse for AppError {
    fn into_response(self) -> Response {
        tracing::error!("AppError sending internal_server_error: {}", self.0);
        let msg = format!("error: {}", self.0);

        (StatusCode::INTERNAL_SERVER_ERROR, msg).into_response()
    }
}

impl<E> From<E> for AppError
    where
        E: Into<anyhow::Error>,
{
    fn from(err: E) -> Self {
        Self(err.into())
    }
}
```

Trouble is, my AppState that I pass into handlers holds a shared reference to both my template engine and the request's cookie jar (so I can identify the user). I'd like to render a pretty error page using these resources - any well documented patterns for this? Basically I want an AppError that can get the same extracts as handler functions, or some other (simple?) pattern.

### axum version

0.7.4

**Comment by jplatte** [comment]
You can use tokio's `task_local!` macro and a middleware that sets the task-local to share state with your `IntoResponse` impl. https://github.com/tokio-rs/axum/pull/2769 has an example with the most important bits for this pattern.
